### ISSUE TITLE
Allow package installation of config package (with yaml files) without the need to have AC Tool package pre-installed

### ISSUE THREAD

**Issue by mickleroy** [Netcentric/accesscontroltool#210]
Big fan of your project - love it!

A pain point I'm running into is the installation of ACTool using the content-package-maven-plugin (I want to avoid installing the package via the UI). 

I've specified the following dependency:
```
<dependencies>
    <dependency>
        <groupId>biz.netcentric.cq.tools.accesscontroltool</groupId>
	<artifactId>accesscontroltool-package</artifactId>
	<version>2.0.1</version>
	<type>zip</type>
	<exclusions>
	    <exclusion>
	        <groupId>biz.netcentric.cq.tools.accesscontroltool</groupId>
	        <artifactId>accesscontroltool-bundle</artifactId>
	    </exclusion>
        </exclusions>
    </dependency>
</dependencies>
```
And the following plugin configuration:
```
<plugin>
    <groupId>com.day.jcr.vault</groupId>
    <artifactId>content-package-maven-plugin</artifactId>
    <extensions>true</extensions>
    <configuration>
        <group>mygroup</group>
        <useProxy>false</useProxy>
        <filterSource>src/main/content/META-INF/vault/filter.xml</filterSource>
        <targetURL>http://${sling.host}:${sling.port}/crx/packmgr/service.jsp</targetURL>
       <properties>
            <installhook.actool.class>biz.netcentric.cq.tools.actool.installhook.AcToolInstallHook</installhook.actool.class>
       </properties>
       <subPackages>
           <subPackage>
                <groupId>biz.netcentric.cq.tools.accesscontroltool</groupId>
                <artifactId>accesscontroltool-package</artifactId>
                <filter>true</filter>
           </subPackage>
      </subPackages>
    </configuration>
</plugin>
```
However, the hook executes before the sub package is installed so the groups/ACLs are not applied. 
```
*INFO* [qtp282918978-3407] org.apache.jackrabbit.vault.packaging.impl.InstallHookProcessorImpl Loading Hook actool: Main-Class = biz.netcentric.cq.tools.actool.installhook.AcToolInstallHook
*ERROR* [qtp282918978-3407] org.apache.jackrabbit.vault.packaging.impl.InstallHookProcessorImpl Error while initializing hook: org.apache.jackrabbit.vault.packaging.PackageException: hook's main class biz.netcentric.cq.tools.actool.installhook.AcToolInstallHook not found: actool
```
Have you attempted to install ACTool this way? What would you recommend?

**Comment by mtstv** [comment]
Yes it is a problem, AEM needs a time after installation of packages, because some service/s are will be re-/started.

I think a configurated timeout must be implemented before the start of Install Hook.

**Comment by mtstv** [comment]
As workaround you must call the next url some time after package installation

`curl -sS --retry 1 -u ${CQ_ADMINUSER}:${CQ_ADMINPW} -X POST "http://${CQ_SERVER}:${CQ_PORT}/system/console/jmx/biz.netcentric.cq.tools:type=ACTool/op/apply/"`

Other thing, you need to install AC Tool only once and not with every software deployment

**Comment by kwin** [comment]
The recommended way is to install the ACTool only once separately. As @mtstv correctly explained you might need to wait some time (until the asynchronous process of deploying the contained OSGi bundle and starting the relevant services has happened). Only afterwards you can successfully install packages leveraging the AC Tool install hook.

**Comment by mtstv** [comment]
> > I think a configurated timeout must be implemented before the start of Install

> Such a thing cannot be implemented in the hook itself, because even that logic would be part of the hook class which is not yet available (classical bootstrapping problem).
So this has to be resolved outside of the hook.

Yes, it is not possible only in case if we install AC Tool package with every our software release.

But in the case if we install AC Tool only once and after that we will deploy our release many times.
The software release usually includes some packages with OSGI configuration as well as a package with AC Tool YAML configurations which needs an install hook. If installation of OSGI configs will trigger the restarting of AEM central services/bundles (I can say it happens in every AEM project) the AEM will break the install hook. To avoid it and if you want to have an automatisation with install hook some configurable pause before starting the install hook will be needed.

Due to this problem we don't use the install hook in my 2 last projects.

**Comment by kwin** [comment]
I opened https://issues.apache.org/jira/browse/JCRVLT-195 about this issue. So hopefully in the future you may use a dependency towards the ACTool bundle in your packages leveraging its install hook.

**Comment by mickleroy** [comment]
Thank you both for your help.

Apologies if this is a silly suggestion (I'm no expert in install hooks) but wouldn't it be possible to have the hook execute on the `Phase.INSTALLED` context instead of `Phase.PREPARE`? This may not work for sub-packages...
https://docs.adobe.com/docs/en/aem/6-2/develop/ref/javadoc/org/apache/jackrabbit/vault/packaging/InstallContext.Phase.html

@mtstv you mentioned you have not used the install hook in your last 2 projects - calling that JMX endpoint (after a timeout) is an alternative?

**Comment by mickleroy** [comment]
That doesn't change the fact you're saying ACTool should only be installed once... This might mean I cannot make use of it for this client as the package installation would have to be done in the AMI baking process.

**Comment by mtstv** [comment]
> @mtstv you mentioned you have not used the install hook in your last 2 projects - calling that JMX endpoint (after a timeout) is an alternative?

@mickleroy Yes, just set a timeout in the deployment shell script after installation of all other packages before calling the 
`curl -sS --retry 1 -u ${CQ_ADMINUSER}:${CQ_ADMINPW} -X POST "http://${CQ_SERVER}:${CQ_PORT}/system/console/jmx/biz.netcentric.cq.tools:type=ACTool/op/apply/"`

**Comment by ghenzler** [comment]
Fo the hen and egg problem we are facing here: potentially we could introduce a new flag in properties.xml that would postpone the actool execution somehow. A simple possibility would be to setup a marker node in JCR somewhere that is picked up by the actool bundle when it's started the first time. the whole mechanism would only kick in for the first installation, all  following  installation runs would work as today.

**Comment by ghenzler** [comment]
Another reason to have this is that for the upgrade case, it is possible to get the following error (if everything is included in one big complete package, I had had this even though the contained packages had correct dependencies defined, but those are only package dependencies and don't help to ensure the sync bundle installation is done)

```
[ERROR] Request failed: org.apache.jackrabbit.vault.packaging.PackageException: hook's main class biz.netcentric.cq.tools.actool.installhook.AcToolInstallHook not found: actool (500)
[WARNING] Deployment of artifact projectGroupId:project-configuration-package:2.16.0-SNAPSHOT:zip failed with exeception: Error while installing package. Check log for details.
```

**Comment by henrykuijpers** [comment]
In #241 I have the issue where, also in the big complete package scenario, the AC Tool bundle would be started after the yaml file has been installed. Consequence: AC tool doesn't "see" the yaml file modification and hence doesn't install acls.

Modifying the yaml file after the automated deployment did trigger the installation of the yaml file. But that should not be needed. 

I was not using the hook, because of the issues that exist with it. However, with the upload listener I ran into a similar timing issue. 

Hopefully we can come to a resolution. In the case of the upload listener, it can simply also check if there is a new (unprocessed) yaml file in the configured location when starting the listener. That would also catch the events that happened in the "past". 

Not sure about the install-hook though...

**Comment by ghenzler** [comment]
To solve this problem, I created https://issues.apache.org/jira/browse/SLING-7790

**Comment by ghenzler** [comment]
As result of https://issues.apache.org/jira/browse/SLING-7790, https://sling.apache.org/documentation/bundles/installer-provider-installhook.html is now available.

**Comment by ghenzler** [comment]
Some more testing revealed that the hook is not perfect yet - waiting for https://issues.apache.org/jira/browse/SLING-8083

**Comment by ghenzler** [comment]
Fixed in v2.2.0

**Comment by brianli-whitbread** [comment]
@ghenzler, maven arch-type 15 and 14 are still broken for both 6.4 and 6.3.
[ERROR] Request failed: org.apache.jackrabbit.vault.packaging.PackageException: Error while executing an install hook during prepare phase. (500)

**Comment by digarg** [comment]
v2.2.0 doesnt seem to have bundle uploaded in repo

**Comment by GR1ML0CK** [comment]
@ghenzler Some people on my team are having the issue described in the original post. I don't seem to be getting this issue. Your last comment on this seems to suggest that this is now working since v2.2.0 (we are using v2.3.2). The code is set up just as the OP's code. Does it need to be implemented differently?
